### PR TITLE
Publish rendered policy artifacts

### PR DESCRIPTION
**Pull request by ggbecker**
#### Description

- Introduces a CMake macro that generates all the rendered policies using as source the control files and publish them in  https://complianceascode.github.io/content-pages/

### CODE REVIEW

**Comment by ggbecker** [comment]
@rhmdnd @jhrozek @Vincent056 

Here's a proof of concept for the rendered policies we talked about. You can find the artifacts here:
https://github.com/ComplianceAsCode/content/suites/13010052155/artifacts/704568550

When we merge, this would be shown in this page: https://complianceascode.github.io/content-pages/

File: .github/workflows/gh-pages.yaml [hunk 1]
```diff
@@ -28,6 +28,9 @@ jobs:
       - name: Build Statistics
         run: ninja html-stats html-profile-stats -j2
         working-directory: ./build
+      - name: Render Policies (Using control files)
+        run: ninja render-policies
```
Comment (ggbecker) [inline .github/workflows/gh-pages.yaml:32]:
```suggestion
        run: ninja render-policies -j2
```

This can be improved by using two parallel jobs

**Comment by marcusburghardt** [comment]
Is it a WIP PR? If so, could you set it as `draft`, please? I would like to confirm based on the title.

**Comment by ggbecker** [comment]
> Is it a WIP PR? If so, could you set it as `draft`, please? I would like to confirm based on the title.

It isn't anymore. I'm keeping the ready status and will update the title.

File: cmake/SSGCommon.cmake [hunk 2]
```diff
@@ -641,6 +641,16 @@ macro(ssg_make_html_stats_for_product PRODUCT)
     )
 endmacro()
 
+macro(ssg_render_policies_for_product PRODUCT CONTROL_FILES)
+        foreach(CONTROL_FILE IN LISTS CONTROL_FILES)
+        add_custom_target(${PRODUCT}-render-policy-${CONTROL_FILE}
```
Comment (yuumasato) [inline cmake/SSGCommon.cmake:646]: I think this should use `add_custom_command` instead, as there is an output file:
https://cmake.org/cmake/help/latest/command/add_custom_target.html

I can run `ninja render-policies` indefinitely and regenerate the artifacts, although there was no change to the DS.

**Review by yuumasato — CHANGES_REQUESTED** [review]

File: utils/generate_html_pages.sh [hunk 3]
```diff
@@ -81,6 +81,37 @@ echo "</body>" >> index.html
 echo "</html>" >> index.html
 popd
 
+
+
+POLICY_DIR="$PAGES_DIR/rendered-policies"
```
[the remark this answers is not in the document]
Comment (ggbecker) [reply utils/generate_html_pages.sh:86]: Ideally a python script, similarly as we do with the guide index.html

https://github.com/ComplianceAsCode/content/blob/master/utils/gen_html_guides_index.py

[the remark this answers is not in the document]
Comment (ggbecker) [reply utils/generate_html_pages.sh:86]: I've moved to a python script and jinja html templates.

File: cmake/SSGCommon.cmake (hunk 2 above)
Comment (ggbecker) [reply cmake/SSGCommon.cmake]: I've switched to `add_custom_command` instead, but it builds nevertheless if it's requested or not. Does it make sense to build this type of content every time we build the project?

Comment (ggbecker) [reply cmake/SSGCommon.cmake]: removing the ALL from the custom target did the trick.

**Comment by ggbecker** [comment]
> The new page looks like this:
> 
> Screenshot from 2023-05-22 14-32-53

the items are now sorted

https://github.com/ComplianceAsCode/content/pull/10585/files#diff-47724278b7434a7e9fc6464470c1be457f5f5c7400d0d2e6b4d49dd95a2d36e1R16

**Comment by ggbecker** [comment]
@yuumasato I think this should be good to go.

File: cmake/SSGCommon.cmake [hunk 4]
```diff
@@ -641,6 +641,20 @@ macro(ssg_make_html_stats_for_product PRODUCT)
     )
 endmacro()
 
+macro(ssg_render_policies_for_product PRODUCT CONTROL_FILES)
+        foreach(CONTROL_FILE IN LISTS CONTROL_FILES)
+        add_custom_command(OUTPUT "${CMAKE_BINARY_DIR}/${PRODUCT}/rendered-policies/${CONTROL_FILE}.html"
```
Comment (yuumasato) [inline cmake/SSGCommon.cmake:646]: Nitpick: move the OUTPUT parameter to the line below.

File: utils/gen_rendered_policies_index.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import pathlib
+import sys
+import yaml
+from collections import namedtuple
+
+import ssg.jinja
+from utils.template_renderer import FlexibleLoader
+from utils.gen_html_guides_index import create_index
+
+# Helper script used to generate an HTML page to display guides.
```
Comment (yuumasato) [inline utils/gen_rendered_policies_index.py:14]: Display control guides? :)

File: cmake/SSGCommon.cmake [hunk 6]
```diff
@@ -775,6 +789,21 @@ macro(ssg_build_product PRODUCT)
     add_dependencies(stats ${PRODUCT}-stats)
     add_dependencies(profile-stats ${PRODUCT}-profile-stats)
     ssg_make_html_stats_for_product(${PRODUCT})
+
+    file(GLOB CONTROL_FILEPATHS "${CMAKE_SOURCE_DIR}/controls/*.yml")
+    message(STATUS "${CONTROL_FILES}")
```
Comment (yuumasato) [inline cmake/SSGCommon.cmake:794]: Does this  `message` have any purpose?

File: utils/gen_rendered_policies_index.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import argparse
+import os
+import pathlib
+import sys
+import yaml
+from collections import namedtuple
+
+import ssg.jinja
+from utils.template_renderer import FlexibleLoader
+from utils.gen_html_guides_index import create_index
```
Comment (yuumasato) [inline utils/gen_rendered_policies_index.py:12]: I value the reuse of code, but this function is not in a module/file in which it is explicit that it is reused.
One update to the `utils/gen_html_guides_index.py` changing that function may break this script.

Comment (ggbecker) [reply utils/gen_rendered_policies_index.py:12]: I agree, it's the codeclimate bothering our lives... I guess I can just return it back be we can waive the codeclimate results.

File: cmake/SSGCommon.cmake (hunk 6 above)
Comment (ggbecker) [reply cmake/SSGCommon.cmake]: Removed, it was for debugging purposes.

File: cmake/SSGCommon.cmake (hunk 4 above)
Comment (ggbecker) [reply cmake/SSGCommon.cmake]: fixed

**Comment by ggbecker** [comment]
@yuumasato feedback addressed

**Review by yuumasato — APPROVED** [review]

**Pull request merged**